src/strategy.rs

```diff
@@ -275,29 +278,225 @@ pub fn hostname<S: BuildHasher>(parameters: Option<HashMap<String, String, S>>)
 }
 
 /// returns true if the strategy should be delegated to, false to disable
-fn _compile_constraint_string<F>(expression: ConstraintExpression, getter: F) -> Evaluate
+fn _compile_constraint_string<F>(
+    expression: ConstraintExpression,
+    inverted: Option<bool>,
+    getter: F,
+) -> Evaluate
 where
     F: Fn(&Context) -> Option<&String> + Clone + Sync + Send + 'static,
 {
     match &expression {
-        ConstraintExpression::In(values) => {
+        ConstraintExpression::In { values } => {
             let as_set: HashSet<String> = values.iter().cloned().collect();
             Box::new(move |context: &Context| {
-                getter(context).map(|v| as_set.contains(v)).unwrap_or(false)
+                let value = getter(context).map(|v| as_set.contains(v)).unwrap_or(false);
+                _handle_inversion(&value, &inverted)
             })
         }
-        ConstraintExpression::NotIn(values) => {
+        ConstraintExpression::NotIn { values } => {
             if values.is_empty() {
-                Box::new(|_| true)
+                Box::new(move |_| _handle_inversion(&true, &inverted))
             } else {
                 let as_set: HashSet<String> = values.iter().cloned().collect();
                 Box::new(move |context: &Context| {
-                    getter(context)
+                    let value = getter(context)
                         .map(|v| !as_set.contains(v))
-                        .unwrap_or(false)
+                        .unwrap_or(false);
+                    _handle_inversion(&value, &inverted)
                 })
             }
         }
+        ConstraintExpression::StrStartsWith {
+            values,
+            case_insensitive,
+        } => {
+            if values.is_empty() {
+                Box::new(move |_| _handle_inversion(&false, &inverted))
+            } else {
+                let mut as_vec: Vec<String> = values.to_vec();
+                let case_insensitive = case_insensitive.unwrap_or(false);
+                if case_insensitive {
+                    as_vec = as_vec.iter_mut().map(|x| x.to_lowercase()).collect();
+                };
+                Box::new(move |context: &Context| {
+                    let value = getter(context)
+                        .map(|v| {
+                            as_vec.iter().any(|x| {
+                                if case_insensitive {
+                                    v.to_lowercase().starts_with(x)
+                                } else {
+                                    v.starts_with(x)
+                                }
+                            })
+                        })
+                        .unwrap_or(false);
+                    _handle_inversion(&value, &inverted)
+                })
+            }
+        }
+        ConstraintExpression::StrEndsWith {
```

Comment:
theres lots of duplication with StrStartsWith: I think a helper function generic over the predicate would be super useful here.